elastic/kibana
# [Discover][Alerting] Use Discover locator for alert results link

## Summary

Closes #145815, #134232

- Moves Discover locator to common area
- Builds alerts results link from the server
- Now there are two implementations of `setStateToKbnUrl` which is used in locator. New one in common are lost `HashedItemStore` support, since sessions storage are actual only for browser
- Toasts `Alert rule has changed`, `Data View has changed` removed
- link generated per each alert will be unique representation of those `rule params` and `data view state` which were at the time of invocation
- Restuls link will live even after data view and rule removal




### How to create rule

- Create an output index and data view `test` 
<details>
  <summary>Query to use</summary>
  
```
PUT test
{
    "settings" : {
        "number_of_shards" : 1
    },
    "mappings" : {
        "properties" : {
            "rule_id" : { "type" : "text" },
            "rule_name" : { "type" : "text" },
            "alert_id" : { "type" : "text" },
            "context_message": { "type" : "text" }
        }
    }
}
```
</details>

- Create alerts connector using `test` index
- Open `Elasticsearch query` alert in `KQL or Lucene` mode or just using Discover `Alerts` button
- Specify the following params: `IS ABOVE: 1`, `FOR THE LAST: 30 min`
- Try execute it by clicking `Test query`. It should match some results
- When choosing connector, use the following config
```
{
    "rule_id": "{{rule.id}}",
    "rule_name": "{{rule.name}}",
    "alert_id": "{{alert.id}}",
    "context_message": "{{context.message}}"
}
```
- Create the alert

### How to test

- Create `Elasticsearch query` rule in `KQL or Lucene` mode like described above
- Wait for some seconds and find the triggered alert document by browsing `test` data view in Discover. There should be a link to results in `context_message` field. Save the link somewhere
- Change rule params by adding/removing filters / changing query / changing data view
- Follow saved link, you should see previous filters, query and data view state
- Open rule in management and click `View in app`, you should see actual state of rule
- Try to remove used data view and then follow saved link, you should still see the results
- Try to remove rule and then follow saved link, you should still see the results.

### Checklist
